### ISSUE TITLE
Amplify Auth | Custom Auth | Amplify crashes if challengeName is DEVICE_PASSWORD_VERIFIER

### ISSUE THREAD

**Issue by Vingtoft** [aws-amplify/amplify-flutter#3541]
### Description

In my application, there are two authentication methods:

A) Username/OTP for first-time users. The OTP is provided via a third-party.
B) Username/Password for returning users on recognized devices.

The below question is in regard to B.

I have created a custom authentication flow, and we start with `PASSWORD_VERIFIER` (checking if username/password is correct) and after that `DEVICE_PASSWORD_VERIFIER` (checking if device is known). 

Please find the "define auth challenge" Lambda trigger below:

```
func Handler(ctx context.Context, event *events.CognitoEventUserPoolsDefineAuthChallenge) (*events.CognitoEventUserPoolsDefineAuthChallenge, error) {
	fmt.Println("Define Auth Challenge! " + strconv.Itoa(len(event.Request.Session)))

	if len(event.Request.Session) == 1 && event.Request.Session[0].ChallengeName == "SRP_A" {
		event.Response.IssueTokens = false
		event.Response.FailAuthentication = false
		event.Response.ChallengeName = "PASSWORD_VERIFIER"
	}

	if len(event.Request.Session) == 2 && event.Request.Session[1].ChallengeName == "PASSWORD_VERIFIER" {
		if event.Request.Session[1].ChallengeResult {
			// username/password was correct, now we check if device is known!
			event.Response.IssueTokens = false
			event.Response.FailAuthentication = false
			event.Response.ChallengeName = "DEVICE_PASSWORD_VERIFIER"
		} else {
			// username/password was incorrect
			event.Response.IssueTokens = false
			event.Response.FailAuthentication = true
		}
	}

	if len(event.Request.Session) == 3 && event.Request.Session[2].ChallengeName == "DEVICE_PASSWORD_VERIFIER" {
		if event.Request.Session[2].ChallengeResult {
			// username/password was correct, AND device is recognized - Auth succeeded!
			event.Response.IssueTokens = true
			event.Response.FailAuthentication = false
		} else {
			// username/password was correct, but device is not recognized - Auth failed!
			event.Response.IssueTokens = false
			event.Response.FailAuthentication = true
		}
	}
	return event, nil
}
```

On the front end, I call the sign In like below:

`var result = await Amplify.Auth.signIn(username: username, password: password);`

The `DEVICE_PASSWORD_VERIFIER` step crashes Cognito with below error:
```

StateMachineManager.accept
state_machine.dart:150
StateMachineManager.acceptAndComplete
state_machine.dart:162
AmplifyAuthCognitoDart.signIn
auth_plugin_impl.dart:536
AuthCategory.signIn
amplify_auth_category.dart:465
_MyHomePageState._signIn
main.dart:106
[VERBOSE-2:dart_vm_initializer.cc(41)] Unhandled Exception: WorkerBeeExceptionImpl {
  error=Null check operator used on a null value,
stackTrace=#0      SrpDevicePasswordVerifierWorker.run
srp_device_password_verifier_worker.dart:86
<asynchronous suspension>
#1      _run
srp_device_password_verifier_worker.worker.vm.dart:14
<asynchronous suspension>
,
}
SrpDevicePasswordVerifierWorker.run
srp_device_password_verifier_worker.dart:86
_run
srp_device_password_verifier_worker.worker.vm.dart:14
===== asynchronous gap ===========================
new EventCompleter
event.dart:43
StateMachineManager.accept
state_machine.dart:150
StateMachineManager.acceptAndComplete
state_machine.dart:162
AmplifyAuthCognitoDart.signIn
auth_plugin_impl.dart:536
AuthCategory.signIn
amplify_auth_category.dart:465
_MyHomePageState._signIn
```

Amplify configuration:

```
{
    "auth": {
      "plugins": {
        "awsCognitoAuthPlugin": {
          "CognitoUserPool": {
            "Default": {
              "PoolId": "eu-west-xx",
              "AppClientId": "xxyyxx",
              "IdentityPoolId": "eu-west-1:xxyyxx",
              "Region": "eu-west-1"
            }
          },
          "CredentialsProvider": {
              "CognitoIdentity": {
                  "Default": {
                      "PoolId": "eu-west-1:xxyyxx",
                      "Region": "eu-west-1"
                  }
              }
          },
          "Auth": {
            "Default": {
              "authenticationFlowType": "CUSTOM_AUTH"
            }
          }
        }
      }
    }
  }
```

**Question:** Why is the frontend crashing when challenge is `DEVICE_PASSWORD_VERIFIER`

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [ ] Authenticator
- [ ] DataStore
- [ ] Notifications (Push)
- [ ] Storage

### Steps to Reproduce

1. Create a custom auth flow like below:

```
func Handler(ctx context.Context, event *events.CognitoEventUserPoolsDefineAuthChallenge) (*events.CognitoEventUserPoolsDefineAuthChallenge, error) {
	fmt.Println("Define Auth Challenge! " + strconv.Itoa(len(event.Request.Session)))

	if len(event.Request.Session) == 1 && event.Request.Session[0].ChallengeName == "SRP_A" {
		event.Response.IssueTokens = false
		event.Response.FailAuthentication = false
		event.Response.ChallengeName = "PASSWORD_VERIFIER"
	}

	if len(event.Request.Session) == 2 && event.Request.Session[1].ChallengeName == "PASSWORD_VERIFIER" {
		if event.Request.Session[1].ChallengeResult {
			// username/password was correct, now we check if device is known!
			event.Response.IssueTokens = false
			event.Response.FailAuthentication = false
			event.Response.ChallengeName = "DEVICE_PASSWORD_VERIFIER"
		} else {
			// username/password was incorrect
			event.Response.IssueTokens = false
			event.Response.FailAuthentication = true
		}
	}

	if len(event.Request.Session) == 3 && event.Request.Session[2].ChallengeName == "DEVICE_PASSWORD_VERIFIER" {
		if event.Request.Session[2].ChallengeResult {
			// username/password was correct, AND device is recognized - Auth succeeded!
			event.Response.IssueTokens = true
			event.Response.FailAuthentication = false
		} else {
			// username/password was correct, but device is not recognized - Auth failed!
			event.Response.IssueTokens = false
			event.Response.FailAuthentication = true
		}
	}
	return event, nil
}
```

2. Sign in `await Amplify.Auth.signIn(username: username, password: password);`

3. Error is displayed

### Screenshots

N/A

### Platforms

- [X] iOS
- [ ] Android
- [X] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.10.6

### Amplify Flutter Version

1.1.0

### Deployment Method

Custom Pipeline

**Comment by dnys1** [comment]
Hi @Vingtoft, I'm working to confirm with the Cognito team, but my understanding is that device SRP is not possible with custom auth. The error you're seeing is due to the fact that Cognito itself is not sending back the parameters necessary for the Flutter library to perform the device SRP handshake.

I will keep you posted while I investigate more.

**Comment by dnys1** [comment]
You mention wanting to understand if the user's device is known. Can you elaborate a bit more on your use case?

**Comment by Vingtoft** [comment]
Thank you @dnys1 !

_"my understanding is that device SRP is not possible with custom auth:"_
As I understand it, device SRP is possible, but only if `SRP_A` and `PASSWORD_VERIFIER` was previously used as challenge. Using custom_auth, the device is actually tracked (I can see the device in the "tracked devices" module in the Cognito console). Isn't it counter intuitive if custom_auth allows you to track device but not to use the device tracking in the auth flow? 

_"You mention wanting to understand if the user's device is known. Can you elaborate a bit more on your use case?:"_
Yes, let me elaborate: 
As a Scandinavian start-up, it's necessary for us to utilize the regions national ID services. In Denmark it's [mitID](https://www.mitid.dk/en-gb/), in Sweden it's [Bank ID](https://www.bankid.com/en/privat/skaffa-bankid). Norway and Finland have similar solutions. 

Using the above solutions creates user trust, but also verifies our users (Know Your Customer (KYC) compliance). 
Asking the user to sign-in using mitId or bankID each time would be a bad user experience. 

We are trying to implement a common solution used by Scandinavian banks:
First time you sign on the app, you must authenticate using mitId (or bankID). You can now chose a 6 digit pin code. 
In the following sessions (on the same device), username + pincode will authenticate you. 

Using a new device, you are required to sign in using the nemID (or bankID). 

So, I need to implement a custom authentication flow: 
1. Does username/password match?
2. Is device tracked? 
Only if 1) and 2) is true, the tokens can be returned. Otherwise, frontend is instructed to start the nemId/bankID flow. 

I hope this answers you questions, please let me know if there's anything I can provide you with.

**Comment by dnys1** [comment]
Hi @Vingtoft, your suspicion was correct. I ended up having to read through the Cognito source to understand how this works, and the answer is not very intuitive 😅 

I believe the Cognito [docs]() and Lambda types are incorrect in that it's not possible to insert device SRP manually into a custom auth flow. Cognito will automatically perform device SRP if the following are true:

1. The user pool has device tracking enabled
2. The user signing in has passed a device key and that device is remembered
3. The current challengeName is `CUSTOM_CHALLENGE` and `issueTokens` is `true`.

For example, the following Lambda code will automatically perform device SRP auth after the custom auth flow is complete (i.e. when `issueTokens = true`):

```ts
import type * as lambda from "aws-lambda";

export const handler: lambda.DefineAuthChallengeTriggerHandler = async (
  event: lambda.DefineAuthChallengeTriggerEvent,
): Promise<lambda.DefineAuthChallengeTriggerEvent> => {
  if (
    event.request.session.length === 1 &&
    event.request.session[0].challengeName === "SRP_A"
  ) {
    event.response.issueTokens = false;
    event.response.failAuthentication = false;
    event.response.challengeName = "PASSWORD_VERIFIER";
  } else if (
    event.request.session.length === 2 &&
    event.request.session[1].challengeName === "PASSWORD_VERIFIER" &&
    event.request.session[1].challengeResult === true
  ) {
    event.response.issueTokens = false;
    event.response.failAuthentication = false;
    event.response.challengeName = "CUSTOM_CHALLENGE";
  } else if (
    event.request.session.length === 3 &&
    event.request.session[2].challengeName === "CUSTOM_CHALLENGE" &&
    event.request.session[2].challengeResult === true
  ) {
    event.response.failAuthentication = false;
    event.response.issueTokens = true;
  } else {
    event.response.issueTokens = false;
    event.response.failAuthentication = true;
  }
  return event;
};
```

The sequence of events from the client's perspective is the following:
- For new logins (i.e. there is no device key) or for logins where the device is tracked but not remembered:
    - `Amplify.Auth.signIn` performs `SRP_A` and `PASSWORD_VERIFIER` challenges and returns `CUSTOM_CHALLENGE` as the next step
    - `Amplify.Auth.confirmSignIn` passes the custom challenge result to the Lambda and, if successful, Cognito issues tokens. The flow is complete.

- For logins where the current device is remembered
	- `Amplify.Auth.signIn` performs `SRP_A` and `PASSWORD_VERIFIER` challenges and returns `CUSTOM_CHALLENGE` as the next step (same as before)
	- `Amplify.Auth.confirmSignIn` passes the custom challenge result to the Lambda and, if successful, Cognito sends `DEVICE_SRP_AUTH` as the next challenge.
	- Internally, the library performs the `DEVICE_SRP_AUTH` and `DEVICE_PASSWORD_VERIFIER` challenges with Cognito (note: your Lambda is no longer invoked for this)
	- If the device challenges are successful, Cognito issues tokens. The flow is complete.

In order for this to work, I do need to make one small change to the library which is passing the `deviceKey` as part of the `confirmSignIn` call. I will have a PR available soon for that and can get it released early next week.

I apologize for the confusion and lack of documentation around this issue. I will confirm with Cognito that I'm not missing anything here and will make sure they're aware of the lack of documentation.

**Comment by Vingtoft** [comment]
Thank you for diving into this issue! 

So, if I understand you correctly, when this change is implemented, I'll have access to "deviceKey" in verifyAuthChallenge (but only if device is tracked and previous challenges was SRP_A, PASSWORD_VERIFYER and "track device" setting is activated in the )

So my auth flow will look like this: 

1. `SRP_A`
2. `VERIFY_PASSWORD`
3. `CUSTOM _CHALLENGE` 
4. Frontend replies custom challenge with Amplify.Auth.confirmSignIn -> now the deviceKey (if any) is appended to the response
5. verifyAuthChallenge can now check if device is remembered by accessing deviceKey.

If the above understanding is (somewhat) correct, then I agree with you that this is a viable solution! 

Please, let me know if I have misunderstood the proposed solution and also let me know a time estimate for when the solution is implemented. 

Thank you so much!

**Comment by dnys1** [comment]
From my investigation, the deviceKey is never passed to your Lambda. This is why I believe the docs are misleading at best. My understanding is that the frontend needs to pass the deviceKey is calls to Cognito and, behind the scenes, Cognito will perform device SRP when your custom auth flow is complete.

**Comment by dnys1** [comment]
I do not believe it is possible to know the status of the device in the custom auth flow.

**Comment by dnys1** [comment]
I've confirmed with Cognito that the device key is never sent to Lambdas and that Cognito alone is responsible for handling device SRP and will do so if the the conditions I mentioned hold true.

I apologize for the poor experience here and have opened a ticket with Cognito to update their docs to highlight all these considerations for device SRP in custom auth flows.

The fix I mentioned is scheduled to be released this week. I'll keep you posted when that happens.

**Comment by dnys1** [comment]
The fix I mentioned has been released in v1.3.3. Thanks for your patience, and please let us know if you continue to face any issues.

Closing this issue for now as I believe all that's left is much better documentation on Cognito's end, which I've brought to their attention and will follow up on async.
